### DISCUSSION TITLE
Bug with default values for Type[T]?

### DISCUSSION THREAD

**Discussion by algmyr** [microsoft/pyright#4545] [General]
In this snippet `f` works perfectly fine, but adding a default value for the type like in `g` makes it fail. I know of workarounds, but am I right this code should work or am I missing something?
```py
from typing import TypeVar, Type

T = TypeVar('T')
def f(t: Type[T]) -> T:
  return t()
def g(t: Type[T]=str) -> T:
  return t()

# Usage-wise, these are meant to be identical.
f(str)
g()
```

**Comment by erictraut** [comment]
Default argument values for parameters annotated with a TypeVar are problematic in the type system. I generally recommend avoiding this if possible. There are a bunch of edge cases that fall apart when they are used. Mypy does not support them at all. Pyright supports them in a limited form, but to be safe, the annotated type is treated as a union of the explicit annotation and the type of the default argument expression. I've considered removing the support entirely and just making it illegal like mypy does. As you point out, there are workarounds (like overloads).

**Reply by algmyr** [reply]
So basically, morally this should work, but implementing it would be very hard to get right? If so I guess that's good enough for me, I mainly wanted to make sure that I wasn't missing anything that would make the code wrong in principle.

**Reply by erictraut** [reply]
> morally this should work, but implementing it would be very hard to get right

I'm not sure what you mean by "morally should work" or "hard to get right". But no, this isn't an issue of code complexity in the type checker. Supporting this case creates holes in the type system that cannot be ignored. I've made it type safe using the technique described above. There are perhaps very simple cases like the one in your example that could be determined to be type safe, but this quickly breaks down in the more general case.

If you're interested in some of the history about this issue, there have been several discussions in the pyright issue tracker, the mypy issue tracker, and the typing-sig mail list. Here are some links:
https://github.com/microsoft/pyright/issues/2507
https://github.com/python/mypy/issues/3737

**Reply by algmyr** [reply]
I just meant that what I wrote is something that "should" work, just that it isn't supported
